cli/azd/pkg/project/framework_service_docker_test.go

```diff
@@ -54,23 +54,27 @@ services:
 	docker := tools.NewDocker(dockerArgs)
 
 	progress := make(chan string)
-	defer close(progress)
+	done := make(chan struct{})
 
 	internalFramework := NewNpmProject(service.Config, &env)
-	status := ""
+	progressMessages := []string{}
 
 	go func() {
 		for value := range progress {
-			status = value
+			progressMessages = append(progressMessages, value)
 		}
+		done <- struct{}{}
 	}()
 
 	framework := NewDockerProject(service.Config, &env, docker, internalFramework)
 	res, err := framework.Package(ctx, progress)
+	close(progress)
+	<-done
```

Comment:
could this be a death lock? 
 for example, if `framework.Package()` never sends a message, would we still be able to exit the blocking for-range on the top to call `done <- struct{}{}` ??

I wonder if we should better use `sync.WaitGroup` and use 2 routines here. One that calls `framerowk.Package` and another one that do the `for-range` to handle messages???

Reply:
No. The `for-range` loop on the channel is terminated by us closing the channel here, not by the number of messages sent. So the case for 0 messages is the same for n. 
1. The goroutine for-range on channel blocks waiting for a message.
2. If a message gets sent, it's processed. If not, it blocks waiting for a message.
3. When the channel is closed (guaranteed here on 71), the goroutine for-range stops iteration.
4. The goroutine sends the `done` message.
5. We wait for the `done` message as a signal that the goroutine is done.
6. We continue execution.